independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/CodestartProcessor.java

```diff
@@ -0,0 +1,159 @@
+
+import io.quarkus.devtools.codestarts.reader.CodestartFile;
+import io.quarkus.devtools.codestarts.reader.CodestartFileReader;
+import io.quarkus.devtools.codestarts.strategy.CodestartFileStrategy;
+import io.quarkus.devtools.codestarts.strategy.CodestartFileStrategyHandler;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+final class CodestartProcessor {
+
+    private final CodestartResourceLoader resourceLoader;
+    private final String languageName;
+    private final Path targetDirectory;
+    private final List<CodestartFileStrategy> strategies;
+    private final Map<String, Object> data;
+    private final Map<String, List<CodestartFile>> files = new LinkedHashMap<>();
+
+    CodestartProcessor(final CodestartResourceLoader resourceLoader,
+            final String languageName,
+            final Path targetDirectory,
+            List<CodestartFileStrategy> strategies,
+            final Map<String, Object> data) {
+        this.resourceLoader = resourceLoader;
+        this.languageName = languageName;
+        this.targetDirectory = targetDirectory;
+        this.strategies = strategies;
+        this.data = data;
+    }
+
+    void process(final Codestart codestart) throws IOException {
+        resourceLoader.loadResourceAsPath(codestart.getResourceDir(), p -> {
+            final Path baseDir = p.resolve(BASE_LANGUAGE);
+            final Path languageDir = p.resolve(languageName);
+            if (!codestart.getSpec().getMissingLanguages().isEmpty()
+                    && codestart.getSpec().getMissingLanguages().contains(languageName)) {
+                // TODO print a warning to inform that this codestart miss this language
+                return null; // We ignore that codestart as it doesn't implement the specified language
+            }
+            Stream.of(baseDir, languageDir)
+                    .filter(Files::isDirectory)
+                    .forEach(dirPath -> processCodestartDir(dirPath,
+                            CodestartData.buildCodestartData(codestart, languageName, data)));
+            return null;
+        });
+    }
+
+    void processCodestartDir(final Path sourceDirectory, final Map<String, Object> finalData) {
+        try {
+            final Collection<Path> sources = Files.walk(sourceDirectory)
+                    .filter(path -> !path.equals(sourceDirectory))
+                    .collect(Collectors.toList());
+            for (Path sourcePath : sources) {
+                final Path relativeSourcePath = sourceDirectory.relativize(sourcePath);
```

Comment:
this code feels like should at least have some debug logging to help to trace/debug issues.

Reply:
yeah that makes sense, I'll add some